yarpl/include/yarpl/flowable/PublishProcessor.h

```diff
@@ -0,0 +1,234 @@
+    }
+  }
+
+  void removePublisher(PublisherSubscription* subscriber) {
+    while (true) {
+      auto oldPublishers = publishers_.copy();
+
+      auto removingItem = std::find_if(
+          oldPublishers->cbegin(),
+          oldPublishers->cend(),
+          [&](const auto& publisherPtr) {
+            return publisherPtr.get() == subscriber;
+          });
+
+      if (removingItem == oldPublishers->cend()) {
+        // not found anymore
+        return;
+      }
+
+      auto newPublishers = std::make_shared<PublishersVector>();
+      newPublishers->reserve(oldPublishers->size() - 1);
+      newPublishers->insert(
+          newPublishers->begin(), oldPublishers->cbegin(), removingItem);
+      newPublishers->insert(
+          newPublishers->end(), std::next(removingItem), oldPublishers->cend());
+
+      auto locked = publishers_.lock();
+      if (*locked == oldPublishers) {
+        *locked = std::move(newPublishers);
+        return;
+      }
+      // else the vector changed so we will have to do it again
+    }
+  }
+
+  class PublisherSubscription : public Subscription {
+   public:
+    PublisherSubscription(
+        std::shared_ptr<Subscriber<T>> subscriber,
+        PublishProcessor* processor)
+        : subscriber_(std::move(subscriber)), processor_(processor) {}
+
+    // cancel may race with terminate(), but the
+    // PublishProcessor::removePublisher will take care of that the race with
+    // on{Next, Error, Complete} methods is allowed by the spec
+    void cancel() override {
+      credits_ = credits::kCanceled;
+      processor_->removePublisher(this);
+    }
+
+    // we don't care about races with this method
+    void request(int64_t n) override {
+      credits::add(&credits_, n);
+    }
+
+    // terminate will never race with on{Next, Error, Complete} because they are
+    // all called from PublishProcessor and terminate is called only from dtor
+    void terminate() {
+      subscriber_->onError(std::runtime_error("PublishProcessor shutdown"));
+    }
+
+    void onNext(T value) {
+      if (credits::tryConsume(&credits_, 1)) {
+        subscriber_->onNext(std::move(value));
+      } else {
+        cancel();
+        subscriber_->onError(MissingBackpressureException());
+      }
+    }
+
+    // used internally, not an interface method
+    void onError(folly::exception_wrapper ex) {
+      subscriber_->onError(std::move(ex));
+    }
+
+    // used internally, not an interface method
+    void onComplete() {
+      subscriber_->onComplete();
+    }
+
+    bool isCancelled() const {
+      return credits_ == credits::kCanceled;
+    }
+
+   private:
+    std::atomic<int64_t> credits_{0};
+    std::shared_ptr<Subscriber<T>> subscriber_;
+    PublishProcessor* processor_;
+  };
+
+  folly::Synchronized<std::shared_ptr<const PublishersVector>, std::mutex>
+      publishers_;
+
+  static std::shared_ptr<const PublishersVector> kCompletedPublishers;
```

Comment:
Do these only exist to "tag" `publishers_`? If so, I would just keep an extra bit of state on `publishers_` within the syncrhonized and test if we're in a `completed` or `errored` state. The magic with the out of line, statically initialized vectors is confusing. Any modules that try to perform static initialization of a `PublishProcessor` are also now subject to the [static initialization order fiasco](https://isocpp.org/wiki/faq/ctors#static-init-order).

Reply:
SIFO can occur if another module attempts to use `PublisherProcessor` directly though. STL types aren't special in the order they're statically initialized.

I think the code will be cleaner if the state is explicit, instead of relying on pointer equality. It'll be easier to introspect the state of the class as well when debugging. There's already a lock (the `folly::Syncronized`) that the state can be put behind.